### PR TITLE
[RNMobile] Latest Posts Block v1 Support in Mobile

### PR DESCRIPTION
**Pull request by chipsnyder**
**Fixes:** https://github.com/wordpress-mobile/gutenberg-mobile/issues/1746
**Related PRs:**
- `gutenberg-mobile` https://github.com/wordpress-mobile/gutenberg-mobile/pull/1931
- For testing:
    - `WordPress-iOS` https://github.com/wordpress-mobile/WordPress-iOS/pull/13495
    - `WordPress-Android` https://github.com/wordpress-mobile/WordPress-Android/pull/11338

## Description
Adds v1 support for the Latest Posts Block on mobile. 

The block in the editor looks similar to an unsupported block but with the word "Customize" instead of unsupported.

After the block is selected you can open the settings menu by:
- Tapping on the block
- Tapping on the gear icon

In the settings, you can configure the following settings:
| Setting | Type | Default | Notes | 
| --- | --- | --- |  --- | 
| Show post content | Toggle | false | |
| Show full post content | Toggle | false | false translates to `excerpt` and true translates to `full_post` | 
| Max number of words in excerpt | range 10 - 100 | 55 | Only available if show post content is on and `Show full post content ` is false |
| Display post date | Toggle | false | |
| Order by | "Newest to Oldest," "Oldest to Newest," "A -> Z," "Z -> A" | Newest to Oldest | |
| Category | All categories configured in site plus a generic one for "All" | All | see "Categories known behaviors" for more info |
| Number of items | range 1 to 100 | 5 | |

## How has this been tested?

These tests were validated via WPiOS and WPAndroid 
Categories make a network request to fetch the available categories, so they were tested against a Wordpress.com site as well as a self-hosted site.

### Add New Block 
1) Select an existing Page or create a new one
2) Add a Latest Posts Block

**Expect:** block to be added

3) Tap on the block or the gear menu

**Expect** the settings to show
**Validate** the block is configured based on the settings in the table above

### Modify Block
This should be tested by modifying in mobile and validating on the web and by modifying on the web and testing on mobile.

1) Select a Latest Post Block and bring up the settings
2) Modify Settings

**Conditional items:** 
- Tapping on the toggle for "Show Post Content" should add a row for "Show" with the options of "Excerpt" or "Full Post" depending on the configuration of the block.
- Changing "Show full post content" to `true` should hide "Max number of words in excerpt" 
- Changing "Show full post content" to `false` should show "Max number of words in excerpt" with the last selected value

3) Close the settings
4) Save as a draft or publish
5) Leave the screen

**Validate** by checking the page on the other platform and by going into the same page/post that all settings match how you configured them.

### Categories known behaviors
Categories are fetched in `componentDidMount` which results in a flow where multiple Latest-Posts Blocks can show different data. 
1) Open a page
2) Add a Latest-Post Block
**- Take Note of the Categories -** 
3) Modify the available categories. If on the web, do this in a new tab.
**- Note the first block does not reflect the changes in the categories -** 
4) Add a new Latest-Post Block
**- Note the second block does reflect the changes in the categories -** 

This behavior is consistent on the web and mobile. Leaving the page and coming back will sync the categories.

## Screenshots 
latestPost

## Types of changes
- Adds v1 support for Latest Post Block

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/latest-posts/edit.native.js
Comment (chipsnyder) [inline packages/block-library/src/latest-posts/edit.native.js]: Android returns a string and iOS returns an object as a result of `fetchRequest`. I debated using a Platform check instead of a type check but thought the type check was safer and more descript.

**Comment by chipsnyder** [comment]
@mkevins and @geriux I assigned both of you to these PRs I was thinking @mkevins could tackle Android and @geriux iOS

File: packages/components/src/index.native.js
Comment (chipsnyder) [inline packages/components/src/index.native.js]: Added this to expose Query Controls to mobile

File: packages/block-library/src/latest-posts/edit.native.js
Comment (mkevins) [reply packages/block-library/src/latest-posts/edit.native.js]: This sounds reasonable. Probably out of scope for this PR, but I'm curious why this platform divergence exists.

I see for Android, we convert the response to a string here:

https://github.com/wordpress-mobile/WordPress-Android/blob/ddb3be8823733e4d6305e63978e8f8353b7c4528/WordPress/src/main/java/org/wordpress/android/ui/posts/reactnative/ReactNativeRequestHandler.kt#L82

and for iOS, we seem to have an object response from here:

https://github.com/wordpress-mobile/WordPressKit-iOS/blob/19a2f186ef6e08e95c4678fd2af487727f3da9a3/WordPressKit/WordPressComRestApi.swift#L203

which is passed along without unwrapping here:

https://github.com/wordpress-mobile/WordPress-iOS/blob/837fbd067d0af8162fa0e21af499a4e469c50b25/WordPress/Classes/ViewRelated/Gutenberg/GutenbergNetworking.swift#L26

cc: @mchowning and @etoledom :wave: :smiley:  Do you know if there is a technical reason the types differ by platform for the `fetchRequest` resolved promise?

Comment (mchowning) [reply packages/block-library/src/latest-posts/edit.native.js]: I can't remember the specific reason Android is using a `String` but there wasn't a technical reason we couldn't pass an object, it was related to it just being much easier to pass the response across the bridge as a String. 

Looking back at it now, I wonder if there is a chance that I wasn't yet aware of the helper methods React Native has for creating `WritableMap`s when I implemented this. I [used one of those helper methods for handling the error response (which I pass across the bridge as an object)](https://github.com/wordpress-mobile/gutenberg-mobile/blob/5ee454d143b5b588038ca11fdc4a90596685e065/react-native-gutenberg-bridge/android/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java#L202), and I can't see why that wouldn't work for the response as well. I think it's worth revisiting whether we can switch Android to passing a successful response across the bridge as an object like we're doing with the error response. I would do that myself, but I really need to focus on the authenticated calls work first, so I created [an issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1961) to track this.

File: packages/components/src/query-controls/index.js
Comment (mkevins) [inline packages/components/src/query-controls/index.js]: This looks familiar :grin: It's nice that the name is the same as that used in Gallery, when the defaults and fencepost logic is sorted out it'll be easier to find and refactor this pattern. :+1:

File: packages/block-library/src/latest-posts/edit.native.js
Comment (chipsnyder) [reply packages/block-library/src/latest-posts/edit.native.js]: @mkevins thanks for doing the leg work on looking up the reason for the difference. I had a todo list to come back to figure out the reasoning behind that, so I appreciate the help there :) 

@mchowning Thanks for the insight! I'll keep an eye on that issue you created. I'll see if I have time to pick that up. It would be good to tackle that soon because it doesn't look like `fetchRequest` is widely used yet.

cc @pinarol

File: packages/block-library/src/latest-posts/edit.native.js
Comment (geriux) [inline packages/block-library/src/latest-posts/edit.native.js]: I think by having all of these within the render function makes it too long.. What do you think if we move these out the `render` and just have them on its own, like:

```
	componentWillUnmount() {
		this.isStillMounted = false;
	}

	onSetExcerptLength = ( value ) => {
		const { setAttributes } = this.props;
		setAttributes( { excerptLength: value } );
	};

	onSetDisplayPostDate = ( value ) => {
		const { setAttributes } = this.props;
		setAttributes( { displayPostDate: value } );
	};

	onSetOrder = ( value ) => {
		const { setAttributes } = this.props;
		setAttributes( { order: value } );
	};

	render() {
```

And then just call them like

```
	<RangeControl
		label={ __( 'Excerpt length (words)' ) }
		value={ excerptLength }
		onChange={ this.onSetExcerptLength }
		min={ MIN_EXCERPT_LENGTH }
		max={ MAX_EXCERPT_LENGTH }
		separatorType="none"
	/>
```

Comment (chipsnyder) [reply packages/block-library/src/latest-posts/edit.native.js]: That works for me 👍

File: packages/block-library/src/index.native.js
Comment (chipsnyder) [reply packages/block-library/src/index.native.js]: That's a good question. I planned to merge without the dev flag as this meets over v1 goals. However, now that you mention it, I'm not sure what the process is for when we wrap in `devOnly` vs. release. 

@pinarol and @iamthomasbishop would you want to release v1 of this or hold off until it's more feature-rich?

Comment (iamthomasbishop) [reply packages/block-library/src/index.native.js]: @chipsnyder Unless I'm missing something, I believe it meets most of our original goals for the v1 so I would be okay with shipping this very basic version. With that said, if we have time to iterate prior to SPTs shipping then maybe we should take advantage of that in the meantime 😀 I'd like to hear @pinarol's thoughts before finalizing, though.

File: packages/block-library/src/latest-posts/edit.native.js
Comment (mkevins) [reply packages/block-library/src/latest-posts/edit.native.js]: Thanks @mchowning for opening that issue and providing the context :+1: .

@chipsnyder glad that helped :smile: . Nice work on this block!

Maybe we can add a small comment as a reminder that this check can be removed once the `fetchRequest` is using the same types on Android and iOS, wdyt?

```suggestion
					// TODO: remove this check after `fetchRequest` types are made consist across platforms
					// (see: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1961)
					if ( typeof categoriesList === 'string' ) {
```

File: packages/block-library/src/latest-posts/edit.native.js
Comment (mkevins) [reply packages/block-library/src/latest-posts/edit.native.js]: Looks good :+1: . Not suggesting any changes here, just wanted to make a minor note that IIRC, web does not use `plugin-proposal-class-properties`, so if we merge `edit.js` and `edit.native.js` in some future iteration, we may need to bind these functions in the constructor.

File: packages/block-library/src/index.native.js
Comment (pinarol) [reply packages/block-library/src/index.native.js]: agreed, this PR covers the goals for v1 and considering @iamthomasbishop gave the 👍 there's no reason to delay it 🎉

**Review by mkevins — COMMENTED** [review]
I tested via the steps in the description, and everything works as described. Nice work :tada: !

#### Minor note (not introduced by this PR):
While testing this, I noticed an issue with the bottomsheet: The `PanelBody` titles do not allow scrolling. Instead, a drag-up does nothing, and a drag-down begins to dismiss the bottomsheet.

#### Screencast:


I'm not sure if this is a known issue?

File: packages/block-library/src/index.native.js
Comment (pinarol) [reply packages/block-library/src/index.native.js]: > With that said, if we have time to iterate prior to SPTs shipping then maybe we should take advantage of that in the meantime

The page template that contains the Latest Posts block isn't included into SPTs 1st iteration, so we don't need to tie shipping SPTs to Latest Posts block. But, we can add that blogger template to SPTs after this PR is merged 🎉
Also, SPTs are ready to go, we just need to remove feature flag.

File: packages/block-library/src/latest-posts/edit.native.js
Comment (chipsnyder) [reply packages/block-library/src/latest-posts/edit.native.js]: Good idea I'll add that comment now.

File: packages/block-library/src/latest-posts/edit.native.js
Comment (chipsnyder) [reply packages/block-library/src/latest-posts/edit.native.js]: I'll have to adjust from this change a bit. I'm getting the following error from the `npm build` script:

`Support for the experimental syntax 'classProperties' isn't currently enabled`

Comment (geriux) [reply packages/block-library/src/latest-posts/edit.native.js]: Yeah @mkevins is right sorry I completely forgot, you'd have to bind them =(

Comment (chipsnyder) [reply packages/block-library/src/latest-posts/edit.native.js]: No worries! I should have those adjusted here shortly testing them now.

**Comment by chipsnyder** [comment]
> I'm not sure if this is a known issue?

Yeah, good call out. I believe this is a known issue because @iamthomasbishop had mentioned it to me in slack. I think this is tracked by this issue https://github.com/wordpress-mobile/gutenberg-mobile/issues/1884

File: packages/block-library/src/latest-posts/edit.native.js
Comment (chipsnyder) [reply packages/block-library/src/latest-posts/edit.native.js]: These changes are good to go now. I didn't recreate an app build for them though so you'll need to use metro to retest.

File: packages/block-library/src/latest-posts/style.native.scss
Comment (mkevins) [inline packages/block-library/src/latest-posts/style.native.scss]: Curious, why do we hard code this in place of the sass var?

File: packages/block-library/src/latest-posts/edit.native.js
Comment (mkevins) [reply packages/block-library/src/latest-posts/edit.native.js]: I re-tested on Android (Pixel 3a) via WordPress-Android:develop and everything is still working as described. :tada:

**Review by geriux — APPROVED** [review]
LGTM! Great work @chipsnyder ! 

Tested on the main iOS app and ✅ 

On the demo app it shows a red screen when opening the settings though, pretty sure its related to the categories but I guess it's not a blocker.

File: packages/block-library/src/latest-posts/style.native.scss
Comment (chipsnyder) [reply packages/block-library/src/latest-posts/style.native.scss]: We had made some design tweaks and I was aligning to what’s in the media placeholder. That used the hex color instead of a var so I really had just copied over settings and didn’t look up to see if that has a matching car defined.

**Comment by chipsnyder** [comment]
> On the demo app it shows a red screen when opening the settings though, pretty sure its related to the categories but I guess it's not a blocker.

That actually sounds odd to me. I didn't test in the demo app but mocked failed network requests and didn't get a red screen. The user should just see "All" in the categories. I'll take a look and make sure there isn't a different object or status code being returned by the demo app that a user might see.

**Comment by chipsnyder** [comment]
> On the demo app it shows a red screen

I'm glad you checked that @geriux! The demo app was returning an object that wasn't an array as a result of its failure where in my other testing the request would fail and fall through to the catch block. So I added some additional type checking to make sure we have the object we expect.

File: packages/block-library/src/latest-posts/edit.native.js
Comment (chipsnyder) [inline packages/block-library/src/latest-posts/edit.native.js]: The demo app was returning an object on its failure to fetch categories. I tried to add `typeof categoriesList === 'array'` here but got a lint error. So instead I'm checking to see if it's empty which seems to handle the weird object case.

**Review by koke — APPROVED** [review]
I didn't look much into the code as this has already one approval, but I tested on Android with a bunch of different combinations of settings, and it seems to be working great 🎉

**Comment by geriux** [comment]
> I'm glad you checked that @geriux! The demo app was returning an object that wasn't an array as a result of its failure where in my other testing the request would fail and fall through to the catch block. So I added some additional checking to make sure we have a non-empty array.

Thank you! Just tested it again on both the demo app and the main app and it's working correctly.

File: packages/block-library/src/latest-posts/style.native.scss
Comment (mkevins) [reply packages/block-library/src/latest-posts/style.native.scss]: Ah, this one was using the `darken` function, so might have been harder to spot: https://github.com/wordpress-mobile/gutenberg-mobile/blob/7500e0eb4d62279659c3af100c1311d255e099a7/src/_colors.scss#L98

File: packages/block-library/src/latest-posts/edit.native.js
Comment (SergioEstevao) [inline packages/block-library/src/latest-posts/edit.native.js]: This shouldn't be used directly and instead, it should use the GB `apiFecth` interface. This will provide support for retries and cache of the data.

The only thing needed is to add the `/wp/v2/categories` to the white list here: https://github.com/wordpress-mobile/gutenberg-mobile/blob/e288c45e882db9808e0396715f379edb07f32416/src/api-fetch-setup.js#L44

@chipsnyder if you need help please ping me or @etoledom.

File: packages/block-library/src/latest-posts/style.native.scss
Comment (chipsnyder) [reply packages/block-library/src/latest-posts/style.native.scss]: Ah yeah that one didn't pop up on my search. I'll make a PR to fix that.

File: packages/block-library/src/latest-posts/edit.native.js
Comment (chipsnyder) [reply packages/block-library/src/latest-posts/edit.native.js]: Thanks @SergioEstevao! I'll make a PR either today or tomorrow morning for that and push it up.

**Pull request merged**
